changefeedccl: kafka sink oauth support

samiskin opened the pull request:
Epic: CRDB-9724

This change adds OAuth support for the Kafka sink via the Client Credentials flow, where the user provides credentials and a url to a token endpoint which CRDB sends requests to to obtain a token it can pass to Sarama to use in its requests to the Kafka broker.


The token endpoint should be standardized between all compliant OAuth servers.  The basic credentials needed to call a Token endpoint are simply a ClientID and ClientSecret, with some optional configuration being possible.  These are sent to a token endpoint either in the POST body or as an authorization Header (some OAuth configurations will accept only one of these formats).  The token endpoint then returns a Token object which contains the token itself as well as an expiry timestamp denoting for how long the token is valid for.  Requests to the Kafka broker are then provided the Token in order to be accepted.

The API for changefeeds is in the form of sink url query parameters just like the existing SASL parameters for kafka sinks:
- `sasl_mechanism=OAUTHBEARER`
- `sasl_client_id=<oauth client id>`
- `sasl_client_secret=<oauth client secret>`
- `sasl_token_url=<full url-encoded url to the token endpoint>`
- Optional: `sasl_scopes=<list of scopes that the token should have access for>` (see [this doc](https://cloud.google.com/apigee/docs/api-platform/security/oauth/working-scopes#:~:text=scopes%20on%20Apigee.-,What%20is%20OAuth2%20scope%3F,resources%2C%20or%20just%20READ%20access.) on scopes)
- Optional: `sasl_grant_type=<custom grant type>` by default this is just `client_credentials` and any compliant Oauth implementation should be using that.  I just added this since the `oauth2/clientcredentials` library allows overriding this value for non-compliant implementations and I figured we might as well support that.


Example CREATE statement for Okta: `CREATE CHANGEFEED FOR foo INTO 'kafka://localhost:9093?sasl_client_id=0oa858flcxdS48fCW5d7&sasl_client_secret=c19KaF9janVndl9JZFdMMFdReTFTbVRyNkdCNnNqdU4xZXZNa19oPQ%3D%3D&sasl_enabled=true&sasl_mechanism=OAUTHBEARER&sasl_token_url=https%3A%2F%2Fdev-13678656.okta.com%2Foauth2%2Fdefault%2Fv1%2Ftoken'`

The way this is done in this PR is the following:
- Credentials for Oauth are passed in as sink query parameters the same as the other forms of auth.  This way we remain consistent with existing methods and they can also be stored in an external connection.  The ClientSecret is expected to be base64 encoded, and the TokenURL must be url encoded.
- We set an AccessTokenProvider implementation for Sarama.  Sarama calls the Token() method on this whenever it attempts to connect to a broker, retrying when errors are observed.
- The Token() method we provide uses the [`oauth2/clientcredentials`](https://pkg.go.dev/golang.org/x/oauth2/clientcredentials#pkg-overview) library, which sends a request to the provided TokenURL, attempting both methods of encoding the credentials (POST body or Header), and caching the knowledge of which one was accepted for future requests. Upon receiving a Token the library also caches the token for future calls until the Expiry timestamp is reached, after which a new token is requested from the endpoint.
- This means that upon a changefeed starting, every node will independently hit the token url with the same credentials, obtain their own token, and then once the expiry timestamp has been reached they will re-request the url.


To verify this works, a roachtest is also added which uses the open source Hydra library to start an in-memory OAuth server.  Using Okta was manually verified with a dev account.


---

The vast majority of the difficulty in getting this working involved setting up OAuth with Kafka for verification purposes.  The main slowdown involved Kafka versions, where OAuth setup between Kafka 2 and Kafka 3 are different, and neither has good documentation around it.

The way it works in Kafka 2 is that a Broker needs to have custom implementations of a Login and a Verification interface that handle connecting to the OAuth provider.  The one used in this PR for the roachtest is from [this
repo](https://github.com/jairsjunior/kafka-oauth).  There were two implementations I found online and both of them were almost identical in source code.  Note that this roachtest setup does not work with Kafka 3. The custom implementation is loaded into Kafka by including the path of the compiled Jar file for it in the CLASSPATH environment variable used when starting Kafka.  The OAuth server parameters used in the custom implementation are also provided via environment variables.

From what I could find, the schema registry does not actually support an OAuth flow when it comes to external clients communicating with the schema registry itself.  The only thing that is supported is OAuth between the schema registry and the SASL-enabled kafka broker.

Release note (enterprise change): changefeeds to a kafka sink now support the OAUTHBEARER sasl_mechanism

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/98053)
<!-- Reviewable:end -->

samiskin committed:
changefeedccl: kafka sink oauth support

Epic: CRDB-9724

This change adds OAuth support for the Kafka sink via the Client
Credentials flow, where the user provides credentials and a url to a
token endpoint which CRDB sends requests to obtain a token it can pass
to Sarama to use in its requests to the Kafka broker.

The token endpoint should be standardized between all compliant OAuth
servers.  The basic credentials needed to call a Token endpoint are
simply a ClientID and ClientSecret, with some optional configuration
being possible.  These are sent to a token endpoint either in the POST
body or as an authorization Header (some OAuth configurations will
accept only one of these formats).  The token endpoint then returns a
Token object which contains the token itself as well as an expiry
timestamp denoting for how long the token is valid for.  Requests to the
Kafka broker are then provided the Token in order to be accepted.

The way this is done in this PR is the following:
- Credentials for Oauth are passed in as sink query parameters the same
  as the other forms of auth.  This way we remain consistent with
  existing methods and they can also be stored in an external
  connection.  The ClientSecret is expected to be base64 encoded as it
  can contain special characters.
- We set an AccessTokenProvider implementation for Sarama.  Sarama calls
  the Token() method on this whenever it attempts to connect to a
  broker, retrying when errors are observed.
- The Token() method we provide uses the oauth2 clientcredentials
  library, which sends a request to the provided TokenURL, attempting
  both methods of encoding the credentials (POST body or Header), and
  caching the knowledge of which one was accepted for future requests.
  Upon receiving a Token the library also caches the token for future
  calls until the Expiry timestamp is reached, after which a new token
  is requested from the endpoint.
- This means that upon a changefeed starting, every node will
  independently hit the token url with the same credentials, obtain
  their own token, and then once the expiry timestamp has been reached
  they will re-request the url.

To verify this works, a roachtest is also added which uses the open
source Hydra library to start an in-memory OAuth server.  Using Okta was
manually verified with a dev account.

---

The vast majority of the difficulty in getting this working involved
setting up OAuth with Kafka for verification purposes.  The main
slowdown involved Kafka versions, where OAuth setup between Kafka 2 and
Kafka 3 are different, and neither has good documentation around it.
The confluent schema registry configuration for OAuth is also different
between version 6.1.0 (which uses kafka 2) and the most recent version,
where it only worked for 6.1.0 between the broker and the schema
registry, not externally.

The way it works in Kafka 2 is that a Broker needs to have custom
implementations of a Login and a Verification interface that handle
connecting to the OAuth provider.  The one used in this PR for the
roachtest is from [this
repo](https://github.com/jairsjunior/kafka-oauth).  There were two
implementations I found online and both of them were almost identical in
source code.  Note that this does not work with Kafka 3. The custom
implementation is loaded into Kafka by including the path of the
compiled Jar file for it in the CLASSPATH environment variable used when
starting Kafka.  The OAuth server parameters used in the custom
implementation are also provided via environment variables.

Release note (enterprise change): changefeeds to a kafka sink now
support the OAUTHBEARER sasl_mechanism

miretskiy reviewed (approved):
Very nice job, @samiskin .  I'm giving this a :thumbsup:  -- from the cdc side, I think things are okay.
Using URL parameters for oath related configs make sense, even if including URL inside url parameter is rather unweildy.

I would like to get confirmation from @renatolabs -- i'm not too comfortable with installing deps in roachtest like you do here, but I don't know if there is a better way to do so anyway.

Finally, do you think it's possible to test this with other oauth providers? Something GCP/AWS specific? I assume those have something that we can use to test.

> Reviewed 4 of 7 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98053#-:-NQ7xhFd-lG9hfqjPjNe:b-v4mn6v)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @HonoreDB, @renatolabs, and @samiskin)

___
*[`-- commits` line 8 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98053#-NQ7uBHN9GBtIk0ltS1b):*
nit: s/to to/to/
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 745 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98053#-NQ7v2el7P2f3Ksllkbe:-NQ7v2el7P2f3Ksllkbf:b-wi1mi7) ([raw file](https://github.com/cockroachdb/cockroach/blob/d92d7de16469d64ca3626a08ede9bad34a7fb99f/pkg/ccl/changefeedccl/sink_kafka.go#L745)):*
> ```Go
> }
> 
> type TokenProvider struct {
> ```

does this need to be exported? If so, probably needs a comment.
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 755 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98053#-NQ7vADMECgrulE0Vx1-:-NQ7vADMECgrulE0Vx10:b-ezlw1m) ([raw file](https://github.com/cockroachdb/cockroach/blob/d92d7de16469d64ca3626a08ede9bad34a7fb99f/pkg/ccl/changefeedccl/sink_kafka.go#L755)):*
> ```Go
> // when connection to the broker fails.
> func (t *TokenProvider) Token() (*sarama.AccessToken, error) {
> 	token, err := t.tokenSource.Token()
> ```

how does sarama handle errors returned from Token()?
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 766 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98053#-NQ7uHQE3P8zda9aKk2x:-NQ7uHQE3P8zda9aKk2y:b-xcwycg) ([raw file](https://github.com/cockroachdb/cockroach/blob/d92d7de16469d64ca3626a08ede9bad34a7fb99f/pkg/ccl/changefeedccl/sink_kafka.go#L766)):*
> ```Go
> 	ctx context.Context, dialConfig kafkaDialConfig,
> ) (sarama.AccessTokenProvider, error) {
> ```

nit: drop \n after {
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 777 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98053#-NQ7vVCEABhFm4BGcaSG:-NQ7vVCEABhFm4BGcaSH:bxwi6hq) ([raw file](https://github.com/cockroachdb/cockroach/blob/d92d7de16469d64ca3626a08ede9bad34a7fb99f/pkg/ccl/changefeedccl/sink_kafka.go#L777)):*
> ```Go
> 	tokenURL, err := url.Parse(dialConfig.saslTokenURL)
> 	if err != nil {
> 		return nil, errors.Wrap(err, "malformed token url")
> ```

let's add a hint that this parameter should be url encoded.
___
*[`pkg/cmd/roachtest/tests/cdc.go` line 1744 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98053#-NQ7xDJe0r-CTP6uRyWU:-NQ7xDJe0r-CTP6uRyWV:bk7ve26) ([raw file](https://github.com/cockroachdb/cockroach/blob/d92d7de16469d64ca3626a08ede9bad34a7fb99f/pkg/cmd/roachtest/tests/cdc.go#L1744)):*
> ```Go
> ) (string, string) {
> 	c.Run(ctx, node, `rm -rf /home/ubuntu/hydra`)
> 	c.Run(ctx, node, `bash <(curl https://raw.githubusercontent.com/ory/meta/master/install.sh) -d -b . hydra v2.0.3`)
> ```

I'm not sure what the word on running installs like this... @renatolabs  do you have any input on  how to best handle dependencies like this in roachtest?


<!-- Sent from Reviewable.io -->

srosenberg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98053)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @HonoreDB, @renatolabs, and @samiskin)

___
*[`pkg/cmd/roachtest/tests/cdc.go` line 1744 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98053#-NQ7xDJe0r-CTP6uRyWU:-NQ8Ys-P43iMySx0zhEz:b-n0v9em) ([raw file](https://github.com/cockroachdb/cockroach/blob/d92d7de16469d64ca3626a08ede9bad34a7fb99f/pkg/cmd/roachtest/tests/cdc.go#L1744)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I'm not sure what the word on running installs like this... @renatolabs  do you have any input on  how to best handle dependencies like this in roachtest?
</blockquote></details>

This will invariably cause occasional flakes. We have many other examples of adhoc install scripts sprinkled over various roachtests. The only way (I know) to combat this is to baked them into a custom AMI. This would eliminate all flakes of this type. However, TestEng would need to provide a way for other teams to edit Packer scripts; e.g., [1]. When we do, we can refactor these. Until then, the best we can do methinks is rely on the built-in retry logic. (Note, you may need to tweak retry options; `DefaultSSHRetryOpts` only retries failed `c.Run` if the error is `ErrSSH255`).

[1] https://github.com/cockroachdb/cockroach/tree/master/build/packer


<!-- Sent from Reviewable.io -->

samiskin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98053)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @HonoreDB, @miretskiy, and @renatolabs)

___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 745 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98053#-NQ7v2el7P2f3Ksllkbe:-NQQ5H-M3gTBOf2GJWHE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d92d7de16469d64ca3626a08ede9bad34a7fb99f/pkg/ccl/changefeedccl/sink_kafka.go#L745)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

does this need to be exported? If so, probably needs a comment.
</blockquote></details>

Done.
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 755 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98053#-NQ7vADMECgrulE0Vx1-:-NQQ4zyUEnt4Dp6UAszm:bphwyo5) ([raw file](https://github.com/cockroachdb/cockroach/blob/d92d7de16469d64ca3626a08ede9bad34a7fb99f/pkg/ccl/changefeedccl/sink_kafka.go#L755)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

how does sarama handle errors returned from Token()?
</blockquote></details>

Sarama will treat it as a broker connection error and retry, logging each time, until it gives up and surfaces the error as a broker connection error.  Added this as a comment.



_Code quote:_
```Go
		// Errors will result in Sarama retrying the broker connection and logging
		// the transient error, with a Broker connection error surfacing after retry
		// attempts have been exhausted.
```
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 777 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98053#-NQ7vVCEABhFm4BGcaSG:-NQQ5HpD-8c_01MqIMpB:b-hwwes9) ([raw file](https://github.com/cockroachdb/cockroach/blob/d92d7de16469d64ca3626a08ede9bad34a7fb99f/pkg/ccl/changefeedccl/sink_kafka.go#L777)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

let's add a hint that this parameter should be url encoded.
</blockquote></details>

Looks like even if I had non-url-encoded strings as tokenURL it didn't trigger this message
___
*[`pkg/cmd/roachtest/tests/cdc.go` line 1744 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98053#-NQ7xDJe0r-CTP6uRyWU:-NQQ5TwnB_bND_XEguU7:b-9dpz33) ([raw file](https://github.com/cockroachdb/cockroach/blob/d92d7de16469d64ca3626a08ede9bad34a7fb99f/pkg/cmd/roachtest/tests/cdc.go#L1744)):*
<details><summary><i>Previously, srosenberg (Stan Rosenberg) wrote…</i></summary><blockquote>

This will invariably cause occasional flakes. We have many other examples of adhoc install scripts sprinkled over various roachtests. The only way (I know) to combat this is to baked them into a custom AMI. This would eliminate all flakes of this type. However, TestEng would need to provide a way for other teams to edit Packer scripts; e.g., [1]. When we do, we can refactor these. Until then, the best we can do methinks is rely on the built-in retry logic. (Note, you may need to tweak retry options; `DefaultSSHRetryOpts` only retries failed `c.Run` if the error is `ErrSSH255`).

[1] https://github.com/cockroachdb/cockroach/tree/master/build/packer
</blockquote></details>

Added retries


<!-- Sent from Reviewable.io -->
